api/resources_portal/views/material_request.py

```diff
@@ -0,0 +1,173 @@
+
+SHARER_MODIFIABLE_FIELDS = {"status", "executed_mta_attachment"}
+
+REQUESTER_MODIFIABLE_FIELDS = {"irb_attachment", "requester_signed_mta_attachment"}
+
+
+class MaterialRequestSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = MaterialRequest
+        fields = (
+            "id",
+            "is_active",
+            "status",
+            "assigned_to",
+            "executed_mta_attachment",
+            "irb_attachment",
+            "material",
+            "requester",
+            "requester_signed_mta_attachment",
+            "created_at",
+            "updated_at",
+        )
+        read_only_fields = ("id", "created_at", "updated_at")
+
+
+class MaterialRequestDetailSerializer(MaterialRequestSerializer):
+    assigned_to = UserSerializer()
+    requester = UserSerializer()
+    material = MaterialSerializer()
+    executed_mta_attachment = AttachmentRelationSerializer()
+    irb_attachment = AttachmentRelationSerializer()
+    requester_signed_mta_attachment = AttachmentRelationSerializer()
+
+
+class CanViewRequests(BasePermission):
+    def has_object_permission(self, request, view, obj):
+        return request.user.has_perm("view_requests", obj.material.organization)
+
+
+class CanViewRequestsOrIsRequester(BasePermission):
+    def has_object_permission(self, request, view, obj):
+        return (
+            request.user.has_perm("view_requests", obj.material.organization)
+            or request.user == obj.requester
+        )
+
+
+class CanApproveRequests(BasePermission):
+    def has_object_permission(self, request, view, obj):
+        return request.user.has_perm("approve_requests", obj.material.organization)
+
+
+class CanApproveRequestsOrIsRequester(BasePermission):
+    def has_object_permission(self, request, view, obj):
+        return (
+            request.user.has_perm("approve_requests", obj.material.organization)
+            or request.user == obj.requester
+        )
+
+
+class IsRequester(BasePermission):
+    def has_object_permission(self, request, view, obj):
+        return request.user == obj.requester
+
+
+class MaterialRequestViewSet(viewsets.ModelViewSet):
+    queryset = MaterialRequest.objects.all()
+
+    def get_serializer_class(self):
+        if self.action == "list":
+            return MaterialRequestSerializer
+
+        return MaterialRequestDetailSerializer
+
+    def get_permissions(self):
+        if self.action == "list":
+            permission_classes = [
+                IsAuthenticated,
+                CanViewRequests,
+            ]
+        elif self.action == "retrieve":
+            permission_classes = [IsAuthenticated, CanViewRequestsOrIsRequester]
+        elif self.action == "update" or self.action == "partial-update":
+            permission_classes = [IsAuthenticated, CanApproveRequestsOrIsRequester]
+        elif self.action == "destroy":
+            permission_classes = [
+                IsAuthenticated,
+                IsRequester,
```

Comment:
Hmm, so it looks like currently we're just allowing the user to delete rather than cancel their requests.

@davidsmejia do you think allowing these to be deleted is good or should we add a `CANCELLED` state that the requester can update the request to? I'm leaning towards having `CANCELLED` but I could maybe be convinced otherwise.

Reply:
I agree with your leanings. We should show cancelled for historical and reporting purposes